[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor

Le-Zheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116004
* #114854
* #114853
* __->__ #114852

**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Le-Zheng committed:
[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor

[ghstack-poisoned]

leslie-fang-intel reviewed (changes requested):
> 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){

@Le-Zheng there are some build errors in the preCI, could you help to check it?  Looks like need to initialize `post_op_algo` with default value.

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -4243,6 +4240,23 @@ def test_qlinear_pt2e(self):
+            'relu': torch.ops.quantized.linear_relu,
+        }
+        supported_post_ops = ['none', 'relu']
+        post_op_algorithm = ''
```
ditto: post_op_algorithms and make it a list.
```suggestion
        post_op_algorithms = []
```

Le-Zheng replied:
Thank you for pointing out and changed accordingly.

jgong5 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -4243,6 +4240,23 @@ def test_qlinear_pt2e(self):
+    def test_qlinear_gelu_pt2e(self):
+        post_op_to_qlinear_ref_dict = {}
+        supported_post_ops = ['gelu']
+        post_op_algorithm = ['none', 'tanh']
```
ditto

Le-Zheng replied:
Thank you for your feedback! I've made the requested changes.

jgong5 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -4188,9 +4181,9 @@ def test_qlinear_pt2e(self):
         cases = itertools.product(
             in_channels_list, out_channels_list, use_bias_list,
-            supported_post_ops, weight_quant_per_channel_list, output_dtype_list)
+            supported_post_ops, weight_quant_per_channel_list, output_dtype_list, post_op_algorithm)
```
Make it plural and make it a list.
```suggestion
            supported_post_ops, weight_quant_per_channel_list, output_dtype_list, post_op_algorithms)
```

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Le-Zheng commented:
> > 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
> > 2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){
> 
> @Le-Zheng there are some build errors in the preCI, could you help to check it? Looks like need to initialize `post_op_algo` with default value.

Thank you for your feedback! I've made the requested changes.

Le-Zheng commented:
> > 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
> > 2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){
> 
> @Le-Zheng there are some build errors in the preCI, could you help to check it? Looks like need to initialize `post_op_algo` with default value.

Thank you for your feedback! I've made the requested changes.

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -967,7 +967,16 @@ static at::Tensor linear_int8_with_onednn_weight(
       tensor::desc(onednn_bias.value().get_dims(), ideep::data_type::f32, ideep::format_tag::any) :
       tensor::desc();
-  auto op_attr = onednn_utils::create_attr_by_post_op(post_op_name, post_op_args);
+  dnnl::algorithm post_op_algo = dnnl::algorithm::eltwise_gelu_erf;
```
nit: I think we should initialize the `post_op_algo` with  `dnnl::algorithm::undef`

Le-Zheng replied:
Thank you for your comment! I've update accordingly.

leslie-fang-intel replied:
I think it should be.... Please take a look again.

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -967,7 +967,16 @@ static at::Tensor linear_int8_with_onednn_weight(
-  auto op_attr = onednn_utils::create_attr_by_post_op(post_op_name, post_op_args);
+  dnnl::algorithm post_op_algo = dnnl::algorithm::eltwise_gelu_erf;
+  if (post_op_name == "gelu"){
+    if (post_op_algorithm == "None"){
```
Should we use `None` or `none`?

Le-Zheng replied:
Thanks for pointing this typo error and changed accordingly.

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor"


**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Le-Zheng commented:
@pytorchbot merge
